
# Project: Future-Group Data Exercise

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

> I will analyze and explore dataset from Future-Group for translation and localization services, this smaple data is an exercise in hiring process for Data Analyst role in the company. 
This data set contains information about sample of 99 observations and including 15 different variables like Account_ID, Account Class, Last Year Growth Indicator, Profitability, Historical and last year Revenues, and other atrractive information to be explored.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
%matplotlib inline
from scipy import stats
import datetime

<a id='wrangling'></a>
## Data Wrangling

In [2]:
import pandas as pd
from pandas import ExcelWriter
from pandas import ExcelFile
df = pd.read_excel('Data Analysis Exercise.xlsx',1)
df.sample(20)

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
51       11165                       LOST          76.79             85.5080   
32       10869                       LOST          82.91         191378.9960   
65       11804                  Increased          48.66           3010.8100   
67       11845                       LOST         -63.79           2678.9200   
84       13238                       LOST         -20.91            161.7700   
43       10918                        NaN            NaN              0.0000   
91       16025                  Decreased          63.93          70140.7247   
75       12306                  Increased          56.29          46094.2099   
35       10880                       LOST          24.55            229.0500   
63       11751                  Increased          62.09         339504.6049   
57       11383                       LOST         100.00            477.1630   
87       13557                       LOST          75.68           1267.3300   
90       15435                        NaN            NaN              0.0000   
22         935                        NaN            NaN                 NaN   
71       11997                       LOST        -999.99            160.0200   
50       11164                  Increased          68.77            682.6500   
58       11402                       LOST          78.02            230.0000   
26        3604                        NaN            NaN              0.0000   
31        8222                       LOST          86.28           8090.0871   
48       10974                       LOST          22.61          21047.6553   

    Last year Revenue  Total open opportunities_USD  \
51                NaN                           NaN   
32                NaN                           0.0   
65                NaN                           NaN   
67                NaN                           NaN   
84                NaN                           NaN   
43                NaN                           NaN   
91          5857.4399                           0.0   
75                NaN                         482.0   
35                NaN                           0.0   
63         13227.7800                      549000.1   
57                NaN                           NaN   
87                NaN                           NaN   
90                NaN                           NaN   
22                NaN                           NaN   
71                NaN                           NaN   
50           682.6500                           NaN   
58                NaN                           NaN   
26                NaN                           NaN   
31                NaN                           NaN   
48                NaN                           NaN   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
51                         1.0                         0.0   
32                      3871.0                         5.0   
65                        34.0                         0.0   
67                        14.0                         0.0   
84                         7.0                         0.0   
43                         1.0                         0.0   
91                       197.0                         0.0   
75                       298.0                         0.0   
35                        11.0                         0.0   
63                      2236.0                         0.0   
57                         2.0                         0.0   
87                        11.0                         0.0   
90                         9.0                         0.0   
22                         NaN                         NaN   
71                        11.0                         0.0   
50                        27.0                         0.0   
58                         1.0                         0.0   
26                        34.0                     

In [3]:
df.shape

(99, 15)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99 entries, 0 to 98
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Account ID                    99 non-null     int64  
 1   Last Year Growth Indicator    79 non-null     object 
 2   Profitability                 79 non-null     float64
 3   Historical Revenue            91 non-null     float64
 4   Last year Revenue             20 non-null     float64
 5   Total open opportunities_USD  27 non-null     float64
 6   Service1OrderCountlastyear    91 non-null     float64
 7   Service2OrderCountlastyear    91 non-null     float64
 8   Service3OrderCountlastyear    91 non-null     float64
 9   Service4OrderCountlastyear    91 non-null     float64
 10  Service5OrderCountlastyear    91 non-null     float64
 11  WeeklyOrderCountlastyear      32 non-null     float64
 12  Monthly OrderCountlastyear    32 non-null     float64
 13  Quarter

- Account ID is string
- Last Year Growth Indicator is category
- Account Class is category

In [5]:
df[df['Last Year Growth Indicator'].isnull()] 

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
3          440                        NaN            NaN                 0.0   
9          524                        NaN            NaN                 0.0   
11         542                        NaN            NaN                 NaN   
12         580                        NaN            NaN                 0.0   
18         718                        NaN            NaN                 NaN   
20         768                        NaN            NaN                 NaN   
22         935                        NaN            NaN                 NaN   
24         999                        NaN            NaN                 NaN   
26        3604                        NaN            NaN                 0.0   
34       10879                        NaN            NaN                 NaN   
36       10881                        NaN            NaN                 0.0   
43       10918                        NaN            NaN                 0.0   
46       10960                        NaN            NaN                 NaN   
47       10970                        NaN            NaN                 0.0   
56       11299                        NaN            NaN                 0.0   
79       12343                        NaN            NaN                 0.0   
89       15041                        NaN            NaN                 0.0   
90       15435                        NaN            NaN                 0.0   
94       18040                        NaN            NaN                 NaN   
95       18049                        NaN            NaN                 0.0   

    Last year Revenue  Total open opportunities_USD  \
3                 NaN                           NaN   
9                 NaN                           NaN   
11                NaN                           NaN   
12                NaN                           NaN   
18                NaN                           NaN   
20                NaN                           NaN   
22                NaN                           NaN   
24                NaN                           0.0   
26                NaN                           NaN   
34                NaN                           NaN   
36                NaN                           NaN   
43                NaN                           NaN   
46                NaN                           NaN   
47                NaN                           NaN   
56                NaN                           NaN   
79                NaN                         100.0   
89                NaN                           NaN   
90                NaN                           NaN   
94                NaN                           NaN   
95                NaN                           NaN   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
3                          6.0                         0.0   
9                        118.0                         1.0   
11                         NaN                         NaN   
12                       119.0                         0.0   
18                         NaN                         NaN   
20                         NaN                         NaN   
22                         NaN                         NaN   
24                         NaN                         NaN   
26                        34.0                         0.0   
34                         NaN                         NaN   
36                         2.0                         0.0   
43                         1.0                         0.0   
46                         NaN                         NaN   
47                        62.0                         0.0   
56                         1.0                         0.0   
79                         0.0                         0.0   
89                         1.0                         0.0   
90                         9.0                     

- Data lack for 20 observations.
- i will deal with the lack of data as i can't assume their values and i can't drop it because the other important variables.

In [6]:
df[df['Historical Revenue'] == 0] 

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
3          440                        NaN            NaN                 0.0   
9          524                        NaN            NaN                 0.0   
12         580                        NaN            NaN                 0.0   
26        3604                        NaN            NaN                 0.0   
36       10881                        NaN            NaN                 0.0   
43       10918                        NaN            NaN                 0.0   
47       10970                        NaN            NaN                 0.0   
56       11299                        NaN            NaN                 0.0   
79       12343                        NaN            NaN                 0.0   
89       15041                        NaN            NaN                 0.0   
90       15435                        NaN            NaN                 0.0   
95       18049                        NaN            NaN                 0.0   

    Last year Revenue  Total open opportunities_USD  \
3                 NaN                           NaN   
9                 NaN                           NaN   
12                NaN                           NaN   
26                NaN                           NaN   
36                NaN                           NaN   
43                NaN                           NaN   
47                NaN                           NaN   
56                NaN                           NaN   
79                NaN                         100.0   
89                NaN                           NaN   
90                NaN                           NaN   
95                NaN                           NaN   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
3                          6.0                         0.0   
9                        118.0                         1.0   
12                       119.0                         0.0   
26                        34.0                         0.0   
36                         2.0                         0.0   
43                         1.0                         0.0   
47                        62.0                         0.0   
56                         1.0                         0.0   
79                         0.0                         0.0   
89                         1.0                         0.0   
90                         9.0                         0.0   
95                         1.0                         0.0   

    Service3OrderCountlastyear  Service4OrderCountlastyear  \
3                          0.0                         0.0   
9                         30.0                         0.0   
12                         0.0                         0.0   
26                         3.0                         0.0   
36                         0.0                         0.0   
43                         0.0                         0.0   
47                         0.0                         0.0   
56                         0.0                         0.0   
79                         0.0                         0.0   
89                         0.0                         0.0   
90                         0.0                         0.0   
95                         0.0                         0.0   

    Service5OrderCountlastyear  WeeklyOrderCountlastyear  \
3                          0.0                       NaN   
9                          0.0                       NaN   
12                         0.0                       NaN   
26                         0.0                       NaN   
36                         0.0                       NaN   
43                         0.0                       NaN   
47                         0.0                       NaN   
56                         0.0                       NaN   
79                         0.0                       NaN   
89                         0.0   

In [7]:
df[df['Account Class'] == 0]

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
0          395                       LOST          72.04            200.3990   
3          440                        NaN            NaN              0.0000   
4          442                       LOST         -12.76         103372.5161   
7          499                       LOST          33.59            160.0000   
10         534                       LOST          71.40          16336.7400   
16         666                       LOST          62.98            349.0510   
19         764                       LOST          62.06           3240.9961   
21         810                       LOST         -11.44            340.2527   
24         999                        NaN            NaN                 NaN   
27        3832                       LOST          39.87           1000.6820   
28        3868                       LOST          76.80             77.9800   
31        8222                       LOST          86.28           8090.0871   
40       10897                       LOST          72.08         147980.8289   
41       10901                       LOST          31.70         169289.4547   
43       10918                        NaN            NaN              0.0000   
45       10950                       LOST          47.13          53676.0125   
48       10974                       LOST          22.61          21047.6553   
49       11157                       LOST          95.32          36706.0000   
53       11211                       LOST          79.84           2590.7230   
54       11247                       LOST          75.70           1001.5200   
55       11288                       LOST          96.21           1407.4400   
56       11299                        NaN            NaN              0.0000   
57       11383                       LOST         100.00            477.1630   
58       11402                       LOST          78.02            230.0000   
60       11423                       LOST          72.01          17528.8509   
61       11573                       LOST          46.12           1085.3590   
64       11793                       LOST         100.00            312.0000   
66       11808                       LOST          83.87           1612.2830   
68       11869                       LOST          56.91            362.6700   
69       11970                       LOST          55.65          15467.6479   
76       12315                       LOST          56.69           4486.4401   
77       12327                       LOST          96.56           6168.6161   
79       12343                        NaN            NaN              0.0000   
81       12711                       LOST          84.74           6381.4350   
82       12914                       LOST          61.00             38.4620   
84       13238                       LOST         -20.91            161.7700   
86       13532                       LOST          78.57           3937.2220   
87       13557                       LOST          75.68           1267.3300   
88       13583                       LOST          94.34            161.9070   
92       18010                       LOST          -9.67            729.4900   
93       18025                       LOST        -999.99             87.1940   
94       18040                        NaN            NaN                 NaN   
95       18049                        NaN            NaN              0.0000   

    Last year Revenue  Total open opportunities_USD  \
0                 NaN                           NaN   
3                 NaN                           NaN   
4                 NaN                           NaN   
7                 NaN                        7560.0   
10                NaN                           NaN   
16                NaN                           NaN   
19                NaN                           NaN   
21                NaN                      

In [8]:
df[df['Account Class'] == 0].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 43 entries, 0 to 95
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Account ID                    43 non-null     int64  
 1   Last Year Growth Indicator    36 non-null     object 
 2   Profitability                 36 non-null     float64
 3   Historical Revenue            41 non-null     float64
 4   Last year Revenue             1 non-null      float64
 5   Total open opportunities_USD  6 non-null      float64
 6   Service1OrderCountlastyear    41 non-null     float64
 7   Service2OrderCountlastyear    41 non-null     float64
 8   Service3OrderCountlastyear    41 non-null     float64
 9   Service4OrderCountlastyear    41 non-null     float64
 10  Service5OrderCountlastyear    41 non-null     float64
 11  WeeklyOrderCountlastyear      1 non-null      float64
 12  Monthly OrderCountlastyear    1 non-null      float64
 13  Quarter

- i have doubts about the records of observations which are nan in Last Year Growth Indicator and Profitability and 0 in Historical Revenue so i preferred to convert their Historical Revenue to nan.
- convert account class = 0 to nan.

In [9]:
df.describe()

Account ID  Profitability  Historical Revenue  Last year Revenue  \
count     99.000000      79.000000        9.100000e+01          20.000000   
mean    8934.949495      32.172025        9.826555e+04        6617.200665   
std     5631.579505     170.097355        2.597567e+05       11757.568498   
min      395.000000    -999.990000        0.000000e+00           0.000000   
25%     1006.000000      49.605000        2.295250e+02         268.905000   
50%    11157.000000      63.090000        3.605616e+03        1848.035000   
75%    12172.000000      76.245000        4.576457e+04        5897.706175   
max    18082.000000     100.000000        1.690395e+06       48241.040500   

       Total open opportunities_USD  Service1OrderCountlastyear  \
count                     27.000000                   91.000000   
mean                   60839.858519                  562.879121   
std                   140664.630377                 1459.870396   
min                        0.000000                    0.000000   
25%                        0.000000                    5.500000   
50%                      100.000000                   27.000000   
75%                     5580.000000                  198.000000   
max                   549000.100000                 8179.000000   

       Service2OrderCountlastyear  Service3OrderCountlastyear  \
count                   91.000000                   91.000000   
mean                     3.054945                   19.780220   
std                     11.558895                   39.855517   
min                      0.000000                    0.000000   
25%                      0.000000                    0.000000   
50%                      0.000000                    1.000000   
75%                      1.000000                   16.000000   
max                    100.000000                  178.000000   

       Service4OrderCountlastyear  Service5OrderCountlastyear  \
count                   91.000000                        91.0   
mean                     0.076923                         0.0   
std                      0.733799                         0.0   
min                      0.000000                         0.0   
25%                      0.000000                         0.0   
50%                      0.000000                         0.0   
75%                      0.000000                         0.0   
max                      7.000000                         0.0   

       WeeklyOrderCountlastyear  Monthly OrderCountlastyear  \
count                 32.000000                   32.000000   
mean                  21.468750                    7.250000   
std                   19.999975                    4.414054   
min                    1.000000                    1.000000   
25%                    4.000000                    3.000000   
50%                   17.500000                    8.000000   
75%                   41.000000                   12.000000   
max                   53.000000                   12.000000   

       QuarterlyOrderCountlastyear  
count                    32.000000  
mean                      2.937500  
std                       1.189673  
min                       1.000000  
25%                       2.000000  
50%                       3.000000  
75%                       4.000000  
max                       4.000000

In [10]:
df.duplicated().sum()

0

In [11]:
df['Account ID'].value_counts(ascending = False).head(5)

768      1
10974    1
11402    1
13557    1
10970    1
Name: Account ID, dtype: int64

In [12]:
df['Profitability'].value_counts(ascending = False).head(5)

-999.99    2
 100.00    2
 83.87     1
 54.09     1
 65.76     1
Name: Profitability, dtype: int64

In [13]:
df[df['Profitability'] == -999.99]

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
71       11997                       LOST        -999.99             160.020   
93       18025                       LOST        -999.99              87.194   

    Last year Revenue  Total open opportunities_USD  \
71                NaN                           NaN   
93                NaN                           NaN   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
71                        11.0                         0.0   
93                        24.0                         0.0   

    Service3OrderCountlastyear  Service4OrderCountlastyear  \
71                         1.0                         0.0   
93                         0.0                         0.0   

    Service5OrderCountlastyear  WeeklyOrderCountlastyear  \
71                         0.0                       NaN   
93                         0.0                       NaN   

    Monthly OrderCountlastyear  QuarterlyOrderCountlastyear Account Class  
71                         NaN                          NaN             C  
93                         NaN                          NaN             0

In [14]:
df[df['Profitability'] == 100]

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
57       11383                       LOST          100.0             477.163   
64       11793                       LOST          100.0             312.000   

    Last year Revenue  Total open opportunities_USD  \
57                NaN                           NaN   
64                NaN                           NaN   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
57                         2.0                         0.0   
64                         2.0                         0.0   

    Service3OrderCountlastyear  Service4OrderCountlastyear  \
57                         0.0                         0.0   
64                         0.0                         0.0   

    Service5OrderCountlastyear  WeeklyOrderCountlastyear  \
57                         0.0                       NaN   
64                         0.0                       NaN   

    Monthly OrderCountlastyear  QuarterlyOrderCountlastyear Account Class  
57                         NaN                          NaN             0  
64                         NaN                          NaN             0

- Profitability = -999.99 or 100 is weird and usual so i preferred to convert it to nan not to drop because it has important observation in historical revenues.

In [15]:
df['Last Year Growth Indicator'].value_counts(ascending = False).head(5)

LOST         48
Increased    18
Decreased    13
Name: Last Year Growth Indicator, dtype: int64

In [16]:
df['Historical Revenue'].value_counts(ascending = False).head(5)

0.0000         12
103372.5161     1
225734.2249     1
32098.6992      1
169289.4547     1
Name: Historical Revenue, dtype: int64

- i have doubts about the records of observations which are nan in Last Year Growth Indicator and Profitability and 0 in Historical Revenue so i preferred to convert their Historical Revenue to nan.

In [17]:
df['Last year Revenue'].value_counts(ascending = False).head(5)

0.0000        1
5416.9300     1
5857.4399     1
13227.7800    1
982.9000      1
Name: Last year Revenue, dtype: int64

In [18]:
df[df['Last year Revenue'] == 0]

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
86       13532                       LOST          78.57            3937.222   

    Last year Revenue  Total open opportunities_USD  \
86                0.0                           NaN   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
86                        12.0                         0.0   

    Service3OrderCountlastyear  Service4OrderCountlastyear  \
86                         2.0                         0.0   

    Service5OrderCountlastyear  WeeklyOrderCountlastyear  \
86                         0.0                       1.0   

    Monthly OrderCountlastyear  QuarterlyOrderCountlastyear Account Class  
86                         1.0                          1.0             0

In [19]:
df['Total open opportunities_USD'].value_counts(ascending = False).head(5)

0.00         13
100.00        2
408000.00     1
3600.00       1
434.08        1
Name: Total open opportunities_USD, dtype: int64

In [20]:
df[df['Total open opportunities_USD'] == 0]

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
24         999                        NaN            NaN                 NaN   
25        1013                  Increased          63.09        2.889747e+04   
28        3868                       LOST          76.80        7.798000e+01   
32       10869                       LOST          82.91        1.913790e+05   
33       10870                  Increased          65.76        1.061761e+06   
35       10880                       LOST          24.55        2.290500e+02   
39       10890                  Increased          62.21        7.701336e+05   
41       10901                       LOST          31.70        1.692895e+05   
52       11174                  Decreased          73.09        7.821665e+04   
70       11979                  Decreased          58.02        2.884511e+05   
78       12334                  Increased          18.06        3.322904e+05   
85       13455                  Increased          53.68        1.120518e+04   
91       16025                  Decreased          63.93        7.014072e+04   

    Last year Revenue  Total open opportunities_USD  \
24                NaN                           0.0   
25         14800.7960                           0.0   
28                NaN                           0.0   
32                NaN                           0.0   
33                NaN                           0.0   
35                NaN                           0.0   
39         25536.9799                           0.0   
41                NaN                           0.0   
52          2289.0500                           0.0   
70          5416.9300                           0.0   
78         48241.0405                           0.0   
85                NaN                           0.0   
91          5857.4399                           0.0   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
24                         NaN                         NaN   
25                        97.0                         0.0   
28                         3.0                         0.0   
32                      3871.0                         5.0   
33                      4900.0                       100.0   
35                        11.0                         0.0   
39                      8179.0                        14.0   
41                      1033.0                         8.0   
52                       507.0                         1.0   
70                      2451.0                         2.0   
78                      2591.0                        11.0   
85                       103.0                         0.0   
91                       197.0                         0.0   

    Service3OrderCountlastyear  Service4OrderCountlastyear  \
24                         NaN                         NaN   
25                        28.0                         0.0   
28                         0.0                         0.0   
32                        65.0                         0.0   
33                        48.0                         0.0   
35                         0.0                         0.0   
39                        90.0                         0.0   
41                       105.0                         0.0   
52                       111.0                         0.0   
70                        11.0                         0.0   
78                        66.0                         7.0   
85                        22.0                         0.0   
91                        10.0                         0.0   

    Service5OrderCountlastyear  WeeklyOrderCountlastyear  \
24                         NaN                       NaN   
25                         0.0                      20.0   
28                         0.0                       NaN   
32                         0.0                       NaN   
33                         0.0                       5.0   
35            

In [21]:
df[df['Total open opportunities_USD'].notnull()] 

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
6          487                  Increased          46.10        3.606546e+05   
7          499                       LOST          33.59        1.600000e+02   
8          520                  Decreased          56.16        1.690395e+06   
14         590                  Decreased          67.06        9.976687e+05   
24         999                        NaN            NaN                 NaN   
25        1013                  Increased          63.09        2.889747e+04   
28        3868                       LOST          76.80        7.798000e+01   
32       10869                       LOST          82.91        1.913790e+05   
33       10870                  Increased          65.76        1.061761e+06   
35       10880                       LOST          24.55        2.290500e+02   
37       10885                  Decreased          61.10        7.578982e+05   
39       10890                  Increased          62.21        7.701336e+05   
41       10901                       LOST          31.70        1.692895e+05   
52       11174                  Decreased          73.09        7.821665e+04   
59       11409                  Decreased          59.14        1.498623e+04   
60       11423                       LOST          72.01        1.752885e+04   
62       11652                       LOST          72.18        2.054000e+02   
63       11751                  Increased          62.09        3.395046e+05   
70       11979                  Decreased          58.02        2.884511e+05   
72       12056                  Increased          54.09        2.257342e+05   
74       12184                  Increased          49.74        4.856900e+02   
75       12306                  Increased          56.29        4.609421e+04   
78       12334                  Increased          18.06        3.322904e+05   
79       12343                        NaN            NaN        0.000000e+00   
85       13455                  Increased          53.68        1.120518e+04   
91       16025                  Decreased          63.93        7.014072e+04   
96       18055                       LOST          81.35        1.379503e+04   

    Last year Revenue  Total open opportunities_USD  \
6           1169.5020                       3000.00   
7                 NaN                       7560.00   
8           1967.4700                     408000.00   
14          6018.5050                     190000.00   
24                NaN                          0.00   
25         14800.7960                          0.00   
28                NaN                          0.00   
32                NaN                          0.00   
33                NaN                          0.00   
35                NaN                          0.00   
37           123.8100                      11175.00   
39         25536.9799                          0.00   
41                NaN                          0.00   
52          2289.0500                          0.00   
59          3745.6000                        100.00   
60                NaN                       3600.00   
62                NaN                        185.00   
63         13227.7800                     549000.10   
70          5416.9300                          0.00   
72          1728.6000                     245040.00   
74                NaN                        434.08   
75                NaN                        482.00   
78         48241.0405                          0.00   
79                NaN                        100.00   
85                NaN                          0.00   
91          5857.4399                          0.00   
96                NaN                     224000.00   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
6                       2279.0                        20.0   
7                         50.0                         0.0   
8                       7248.0      

In [22]:
df[df['Total open opportunities_USD'] == 100]

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
59       11409                  Decreased          59.14           14986.227   
79       12343                        NaN            NaN               0.000   

    Last year Revenue  Total open opportunities_USD  \
59             3745.6                         100.0   
79                NaN                         100.0   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
59                       109.0                         2.0   
79                         0.0                         0.0   

    Service3OrderCountlastyear  Service4OrderCountlastyear  \
59                         6.0                         0.0   
79                         0.0                         0.0   

    Service5OrderCountlastyear  WeeklyOrderCountlastyear  \
59                         0.0                       5.0   
79                         0.0                       NaN   

    Monthly OrderCountlastyear  QuarterlyOrderCountlastyear Account Class  
59                         4.0                          3.0             C  
79                         NaN                          NaN             0

- convert Total open opportunities_USD = 0  to nan as it unlogic.
- drop the account id = 12343 as it has not data to be explored.

In [23]:
df['QuarterlyOrderCountlastyear'].value_counts(ascending = False).head(5)

4.0    15
3.0     6
1.0     6
2.0     5
Name: QuarterlyOrderCountlastyear, dtype: int64

In [24]:
df['Service1OrderCountlastyear'].sort_values(ascending = False).head(10)

39    8179.0
8     7248.0
14    5027.0
33    4900.0
32    3871.0
78    2591.0
70    2451.0
37    2379.0
6     2279.0
63    2236.0
Name: Service1OrderCountlastyear, dtype: float64

### Data Cleaning

### To be cleaned
- Account ID is string.
- Last Year Growth Indicator is category.
- Account Class is category.
- convert account class = 0 to nan.
- Profitability = -999.99 or 100 is weird and usual so i preferred to convert it to nan not to drop because it has important observation in historical revenues.
- i have doubts about the records of observations which are nan in Last Year Growth Indicator and Profitability and 0 in Historical Revenue so i preferred to convert their Historical Revenue to nan.
- convert Total open opportunities_USD = 0  to nan as it unlogic.
- drop the account id = 12343 as it has not data to be explored.

In [25]:
df_clean = df.copy()

In [26]:
# Account ID is string.
df_clean['Account ID'] = df_clean['Account ID'].astype('str')

# Last Year Growth Indicator is category.
df_clean['Last Year Growth Indicator'] = df_clean['Last Year Growth Indicator'].astype('category')

9# Account Class is category.
df_clean['Account Class'] = df_clean['Account Class'].astype('category')

df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99 entries, 0 to 98
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype   
---  ------                        --------------  -----   
 0   Account ID                    99 non-null     object  
 1   Last Year Growth Indicator    79 non-null     category
 2   Profitability                 79 non-null     float64 
 3   Historical Revenue            91 non-null     float64 
 4   Last year Revenue             20 non-null     float64 
 5   Total open opportunities_USD  27 non-null     float64 
 6   Service1OrderCountlastyear    91 non-null     float64 
 7   Service2OrderCountlastyear    91 non-null     float64 
 8   Service3OrderCountlastyear    91 non-null     float64 
 9   Service4OrderCountlastyear    91 non-null     float64 
 10  Service5OrderCountlastyear    91 non-null     float64 
 11  WeeklyOrderCountlastyear      32 non-null     float64 
 12  Monthly OrderCountlastyear    32 non-null     float6

In [27]:
# convert account class = 0 to nan.
df_clean.loc[df_clean['Account Class'] == 0, 'Account Class'] = np.nan

# convert Profitability = -999.99 or 100 to nan.
df_clean.loc[df_clean['Profitability'] == 100, 'Profitability'] = np.nan
df_clean.loc[df_clean['Profitability'] == -999.99, 'Profitability'] = np.nan

# convert Historical Revenue = 0 to nan.
df_clean.loc[df_clean['Historical Revenue'] == 0, 'Historical Revenue'] = np.nan

# convert Total open opportunities_USD = 0  to nan.
df_clean.loc[df_clean['Total open opportunities_USD'] == 0, 'Total open opportunities_USD'] = np.nan

# drop the account id = 12343 as it has not data to be explored.
df_clean.drop(df_clean.index[79], axis=0, inplace = True)

In [28]:
df_clean.sample(20)

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
80      12695                  Decreased         -31.12         148904.4237   
11        542                        NaN            NaN                 NaN   
12        580                        NaN            NaN                 NaN   
1         405                       LOST          94.21           3605.6160   
0         395                       LOST          72.04            200.3990   
62      11652                       LOST          72.18            205.4000   
95      18049                        NaN            NaN                 NaN   
76      12315                       LOST          56.69           4486.4401   
66      11808                       LOST          83.87           1612.2830   
86      13532                       LOST          78.57           3937.2220   
27       3832                       LOST          39.87           1000.6820   
44      10943                  Increased          71.23           1234.5130   
92      18010                       LOST          -9.67            729.4900   
16        666                       LOST          62.98            349.0510   
14        590                  Decreased          67.06         997668.7253   
75      12306                  Increased          56.29          46094.2099   
29       4112                       LOST          39.69           1055.2500   
9         524                        NaN            NaN                 NaN   
13        586                  Increased          49.47          23273.6632   
38      10888                       LOST          55.10          11335.4320   

    Last year Revenue  Total open opportunities_USD  \
80            317.270                           NaN   
11                NaN                           NaN   
12                NaN                           NaN   
1                 NaN                           NaN   
0                 NaN                           NaN   
62                NaN                         185.0   
95                NaN                           NaN   
76                NaN                           NaN   
66                NaN                           NaN   
86              0.000                           NaN   
27                NaN                           NaN   
44                NaN                           NaN   
92                NaN                           NaN   
16                NaN                           NaN   
14           6018.505                      190000.0   
75                NaN                         482.0   
29                NaN                           NaN   
9                 NaN                           NaN   
13                NaN                           NaN   
38                NaN                           NaN   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
80                       343.0                         0.0   
11                         NaN                         NaN   
12                       119.0                         0.0   
1                         66.0                         0.0   
0                          5.0                         0.0   
62                         2.0                         0.0   
95                         1.0                         0.0   
76                         7.0                         0.0   
66                        25.0                         0.0   
86                        12.0                         0.0   
27                        15.0                         1.0   
44                         5.0                         0.0   
92                         3.0                         0.0   
16                         1.0                         0.0   
14                      5027.0                        19.0   
75                       298.0                         0.0   
29                         3.0                         2.0   
9                        118.0                         1.0   
13       

In [29]:
df_clean[df_clean['Account ID'] == '12343']

Empty DataFrame
Columns: [Account ID, Last Year Growth Indicator, Profitability, Historical Revenue, Last year Revenue, Total open opportunities_USD, Service1OrderCountlastyear, Service2OrderCountlastyear, Service3OrderCountlastyear, Service4OrderCountlastyear, Service5OrderCountlastyear, WeeklyOrderCountlastyear, Monthly OrderCountlastyear, QuarterlyOrderCountlastyear, Account Class]
Index: []

In [30]:
df_clean.sample(20)

Account ID Last Year Growth Indicator  Profitability  Historical Revenue  \
14        590                  Decreased          67.06        9.976687e+05   
46      10960                        NaN            NaN                 NaN   
70      11979                  Decreased          58.02        2.884511e+05   
81      12711                       LOST          84.74        6.381435e+03   
12        580                        NaN            NaN                 NaN   
8         520                  Decreased          56.16        1.690395e+06   
74      12184                  Increased          49.74        4.856900e+02   
78      12334                  Increased          18.06        3.322904e+05   
29       4112                       LOST          39.69        1.055250e+03   
91      16025                  Decreased          63.93        7.014072e+04   
1         405                       LOST          94.21        3.605616e+03   
21        810                       LOST         -11.44        3.402527e+02   
85      13455                  Increased          53.68        1.120518e+04   
40      10897                       LOST          72.08        1.479808e+05   
86      13532                       LOST          78.57        3.937222e+03   
94      18040                        NaN            NaN                 NaN   
54      11247                       LOST          75.70        1.001520e+03   
52      11174                  Decreased          73.09        7.821665e+04   
89      15041                        NaN            NaN                 NaN   
36      10881                        NaN            NaN                 NaN   

    Last year Revenue  Total open opportunities_USD  \
14          6018.5050                     190000.00   
46                NaN                           NaN   
70          5416.9300                           NaN   
81                NaN                           NaN   
12                NaN                           NaN   
8           1967.4700                     408000.00   
74                NaN                        434.08   
78         48241.0405                           NaN   
29                NaN                           NaN   
91          5857.4399                           NaN   
1                 NaN                           NaN   
21                NaN                           NaN   
85                NaN                           NaN   
40                NaN                           NaN   
86             0.0000                           NaN   
94                NaN                           NaN   
54                NaN                           NaN   
52          2289.0500                           NaN   
89                NaN                           NaN   
36                NaN                           NaN   

    Service1OrderCountlastyear  Service2OrderCountlastyear  \
14                      5027.0                        19.0   
46                         NaN                         NaN   
70                      2451.0                         2.0   
81                        51.0                         0.0   
12                       119.0                         0.0   
8                       7248.0                        33.0   
74                        10.0                         0.0   
78                      2591.0                        11.0   
29                         3.0                         2.0   
91                       197.0                         0.0   
1                         66.0                         0.0   
21                         2.0                         0.0   
85                       103.0                         0.0   
40                      2079.0                         2.0   
86                        12.0                         0.0   
94                         NaN                         NaN   
54                         4.0                         0.0   
52                       507.0                         1.0   
89       

In [31]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 98 entries, 0 to 98
Data columns (total 15 columns):
 #   Column                        Non-Null Count  Dtype   
---  ------                        --------------  -----   
 0   Account ID                    98 non-null     object  
 1   Last Year Growth Indicator    79 non-null     category
 2   Profitability                 75 non-null     float64 
 3   Historical Revenue            79 non-null     float64 
 4   Last year Revenue             20 non-null     float64 
 5   Total open opportunities_USD  13 non-null     float64 
 6   Service1OrderCountlastyear    90 non-null     float64 
 7   Service2OrderCountlastyear    90 non-null     float64 
 8   Service3OrderCountlastyear    90 non-null     float64 
 9   Service4OrderCountlastyear    90 non-null     float64 
 10  Service5OrderCountlastyear    90 non-null     float64 
 11  WeeklyOrderCountlastyear      32 non-null     float64 
 12  Monthly OrderCountlastyear    32 non-null     float6

In [32]:
df_clean.describe()

Profitability  Historical Revenue  Last year Revenue  \
count      75.000000        7.900000e+01          20.000000   
mean       57.887600        1.131920e+05        6617.200665   
std        30.149991        2.759400e+05       11757.568498   
min       -63.790000        3.846200e+01           0.000000   
25%        51.710000        7.060700e+02         268.905000   
50%        63.090000        8.090087e+03        1848.035000   
75%        75.690000        6.700459e+04        5897.706175   
max        96.560000        1.690395e+06       48241.040500   

       Total open opportunities_USD  Service1OrderCountlastyear  \
count                     13.000000                   90.000000   
mean                  126352.013846                  569.133333   
std                   184130.244293                 1466.822555   
min                      100.000000                    1.000000   
25%                      482.000000                    6.250000   
50%                     7560.000000                   28.000000   
75%                   224000.000000                  198.500000   
max                   549000.100000                 8179.000000   

       Service2OrderCountlastyear  Service3OrderCountlastyear  \
count                   90.000000                   90.000000   
mean                     3.088889                   20.000000   
std                     11.619090                   40.023308   
min                      0.000000                    0.000000   
25%                      0.000000                    0.000000   
50%                      0.000000                    1.000000   
75%                      1.000000                   17.000000   
max                    100.000000                  178.000000   

       Service4OrderCountlastyear  Service5OrderCountlastyear  \
count                   90.000000                        90.0   
mean                     0.077778                         0.0   
std                      0.737865                         0.0   
min                      0.000000                         0.0   
25%                      0.000000                         0.0   
50%                      0.000000                         0.0   
75%                      0.000000                         0.0   
max                      7.000000                         0.0   

       WeeklyOrderCountlastyear  Monthly OrderCountlastyear  \
count                 32.000000                   32.000000   
mean                  21.468750                    7.250000   
std                   19.999975                    4.414054   
min                    1.000000                    1.000000   
25%                    4.000000                    3.000000   
50%                   17.500000                    8.000000   
75%                   41.000000                   12.000000   
max                   53.000000                   12.000000   

       QuarterlyOrderCountlastyear  
count                    32.000000  
mean                      2.937500  
std                       1.189673  
min                       1.000000  
25%                       2.000000  
50%                       3.000000  
75%                       4.000000  
max                       4.000000

In [33]:
# storing cleaned data to excel sheet.

df_clean.to_excel("Cleaned_Data.xlsx")

<a id='conclusions'></a>

## Conclusions

#### After loading data from XLSX file of Future Group Exercise Dataset, i started assessing data visually and programatically to be cleaned and I found some limitations i need to handle it before starting analysis , for instance:

- Convert Account ID to string.
- Convert Last Year Growth Indicator to category.
- Convert Account Class to category.
- convert account class = 0 to nan.
- Profitability = -999.99 or 100 is weird and usual so i preferred to convert it to nan not to drop because it has important observation in historical revenues.
- I have doubts about the records of observations which are nan in Last Year Growth Indicator and Profitability and 0 in Historical Revenue so i preferred to convert their Historical Revenue to nan.
- Convert Total open opportunities_USD = 0  to nan as it unlogic.
- Drop the account id = 12343 as it has not data to be explored.

#### I decided to export the cleaned data to tableau where I can do the explanatory process and I got great insights and findings. You can find Tableau dashboards through this link: https://public.tableau.com/app/profile/khaled.yaseen/viz/Future-GroupExercise/AccountClassGrowthindicator?publish=yes. 
#### and here are the insights I found:


- The Account with ID: 520 is the most important customer and has a huge gap from other accounts. Its historical revenues ~= 1700000, In the last year company gained from it ~= 2000 and it is small number comparing to other top accounts and its historical revenues so its Last Year Growth Indicator decreased.

- The Account with ID: 12334 is the first account in the last year revenues and has revenues ~= 49000 with a huge gap from the second account with ID: 10890 which have revenues ~= 26000. And this is only its first year with the company so it is a good start to grow investment with it.

- The Account with ID: 10890 has the most service 2 order count in the last year with 8179 orders then the account ID: 520( Historically the most important customer) with 7248 orders.

- The Account with ID: 10870 has the most service 2 order count in the last year with 100 orders then the account ID: 520 -Historically the most important customer- with 33 orders.

- The Account with ID: 10885 has the most service 3 order count in the last year with 178 orders then the account ID: 11751 with 159 orders. Note: the two accounts are CLASS B.

- Only account with ID: 12334 -the first account in the last year revenues- has service 4 order count in last year with 7 orders.

- There is no accounts has service 5 order count in the last year.

- The highest and the lowest accounts in profitability are in the lost category in the last year growth indicator and have low revenues but in general the lost last year growth indicator is the smallest in avg last year growth profitability. the highest account is: 12327 with profitability = 96.4 and the lowest is: 11845 with profitability = -63.8.

- The highest account in total open opportunities is: 11751 with total open opportunities = 549000.

- Class C is the most class represented in data with percentile 52.73% of total classes and has the highest avg profitability with 62.66.

- Class A is the highest class in historical revenues with percentile 64.14% of total historical revenues for classes.

- Last year growth indicator: Lost is the highest indicator in data with percentile 60.76% of total indicators so company should solve this problem to keep customers.

- Last year growth indicator: decreased is the highest indicator in historical revenues percentiles with percentile 52.49% of total indicators so company should do its best to make the highest indicator: increased.

- Last year growth indicator: increased has the highest avg profitability slightly different to other indicators. incresed = 59.3 , decreased = 58.7 , lost = 57.0.
  
- Account class: C and Last year growth indicator: decreased is the most case occur but has small revenues comparing to class: A Whether it increases or decreases in Last year growth indicator but But surprisingly, this account class: C and Last year growth indicator: decreased is higher than all cases for class: B.

- Service 1&2&3 order counts has a positive relationship with revenues.

- Monthly and quarterly order counts for Account class: A,  almost has a positive relation with revenues. and for weekly order count it is a positive relationship but the revenues reduce from weekly order count 49 to 51. and account classes: B&C also have a positive relation with revenues.

- Profitability has a small negative relationship with revenues and that is strange for me.

- I plotted a chart between service 1&2&3 order counts for easy predictions between each other.